### PR TITLE
KUI-1257 Improve client-side language handling

### PR DESCRIPTION
**Pull request by belanglos**
Throughout our projects we frequently encounter duplicated language-related checks,
e.g.
* `lang === 'en' ? 'something' : 'other'`
* i18n.messages[languageIndex]
* const doSomething = (language = 'sv')

This PR is an attempt at getting rid of as much duplicated code as possible and centralize checks, default values and language handling in general as much as possible.

A couple of hooks have been introduced, amongst others:
* `useLanguage`, which should work as "source of truth" for everything language-related
* `useMissingInfo`, which is an attempt at getting rid of in-line comparisons with `INFORM_IF_IMPORTANT_INFO_IS_MISSING`. This is not the best solution, but hopefully a step on the way. 

In some apps (albeit not this one), we also use multiple languages for different parts of the UI at the same time. The goal is to adapt the `useLanguage` hook to those apps as well.

Note that this PR only changes client-side language-handling.

### CODE REVIEW

File: public/js/app/hooks/useRoundUtils.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import React from 'react'
+import { useLanguage } from './useLanguage'
+import { useMissingInfo } from './useMissingInfo'
+
+export const useRoundUtils = () => {
```
Comment (karlandindrakryggen) [inline public/js/app/hooks/useRoundUtils.js:5]: Maybe `useRoundLabel` `useRoundHeader` instead of `useRoundUtils`? Lets discuss this

Comment (karlandindrakryggen) [reply public/js/app/hooks/useRoundUtils.js:5]:
```
export const useRoundLabel = round => {
  const { translation } = useLanguage()
  const { isMissingInfoLabel } = useMissingInfo()

  const { round_short_name, round_funding_type, round_category } = round
  let fundingType
  if (round_funding_type === 'PER' || round_funding_type === 'UPP' || round_funding_type === 'SAP') {
    fundingType = translation.courseRoundInformation.round_type[round_funding_type]
  } else {
    fundingType = translation.courseRoundInformation.round_category[round_category]
  }
  return ` ${!isMissingInfoLabel(round_short_name) ? `${round_short_name}` : ''} ${fundingType}`
}

```

**Review by karlandindrakryggen — APPROVED** [review]
Nice job!

Just fix the comment in SideMenu test.

File: public/js/app/components/SideMenu.jsx [hunk 2]
```diff
@@ -19,22 +32,25 @@ const labelBeforeChoosingCourse = (courseCode, label) =>
     </p>
   ) : null
 
-const SideMenu = ({ courseCode, labels = {}, language }) => {
-  const courseDevelopmentLink = `${COURSE_DEVELOPMENT_URL}${courseCode}?l=${language}`
-  const courseArchiveLink = `${COURSE_DEVELOPMENT_URL}${courseCode}/arkiv?l=${language}`
+const SideMenu = ({ courseCode, labels = {} }) => {
+  const { languageShortname, isEnglish } = useLanguage()
+
+  const { aboutCourseLink, courseMemoLink } = createLinksWithOptionalLanguageParameter(courseCode, isEnglish)
```
[the remark this answers is not in the document]
Comment (belanglos) [reply public/js/app/components/SideMenu.jsx:38]: Very good question. As far as I understand it, if we do not send the language parameter, we let the receiving application decide which language to choose. I don't know whether this is intentional, but as it is coded this way, I hope it is.

File: public/js/app/components/__tests__/SideMenu.test.js [hunk 3]
```diff
@@ -34,7 +39,11 @@ describe('Component <SideMenu>', () => {
   })
 
   test('renders a side menu for a course and with labels. Swedish', () => {
-    render(<SideMenu courseCode={'KIP1111'} labels={translationSV.courseLabels.sideMenu} language={'sv'} />)
+    render(
+      <WebContextProvider contconfigInextIn={{ lang: 'sv' }}>
```
[the remark this answers is not in the document]
Comment (belanglos) [reply public/js/app/components/__tests__/SideMenu.test.js:43]: Good catch! Will get back to that!

File: public/js/app/hooks/useFormatCredits.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    creditUnitAbbr => (isEnglish ? 'credits' : creditUnitAbbr),
+    [isEnglish]
+  )
+
+  const ensureCreditIsNumber = credits => +credits
+
+  const formatCredits = React.useCallback(
+    (credits, creditUnitAbbr) => {
+      // I want to use global isNaN as it works differently
+      // eslint-disable-next-line no-restricted-globals
```
[the remark this answers is not in the document]
Comment (belanglos) [reply public/js/app/hooks/useFormatCredits.js:30]: I learned it just a few minutes before writing that comment ;-)

File: public/js/app/components/SideMenu.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (karlandindrakryggen) [reply public/js/app/components/SideMenu.jsx:38]: Maybe something to look more into in the future, ie why it is different and maybe change it to work in same way in all apps.

**Pull request merged**